src/components/flowchart/flowchart.js

```diff
@@ -451,23 +491,66 @@ export class FlowChart extends Component {
    */
   handleNodeClick = (event, node) => {
     const { type, id } = node;
-    const {
-      onClickToExpandModularPipeline,
-      displayMetadataPanel,
-      onLoadNodeData,
-      toSelectedNode,
-      onToggleNodeClicked,
-    } = this.props;
+    const { onClickToExpandModularPipeline } = this.props;
 
     if (type === 'modularPipeline') {
       onClickToExpandModularPipeline(id);
     } else {
-      displayMetadataPanel ? onLoadNodeData(id) : onToggleNodeClicked(id);
-      toSelectedNode(node);
+      this.handleSingleNodeClick(node);
+
+      // the hold shift only happens on clicking a node first
+      if (event.shiftKey) {
+        this.handleShiftClick(node);
+      }
     }
+
     event.stopPropagation();
   };
 
+  handleSingleNodeClick = (node) => {
+    const { id } = node;
+    const {
+      displayMetadataPanel,
+      onLoadNodeData,
+      onToggleNodeClicked,
+      toSelectedNode,
+    } = this.props;
+
+    displayMetadataPanel ? onLoadNodeData(id) : onToggleNodeClicked(id);
+    toSelectedNode(node);
+
+    const { from, to } = this.state.filteredPipelineState;
+    // Reset or set the first node as the 'from' node based on current state
+    const newState =
+      from !== null && to !== null
+        ? { ...this.state.filteredPipelineState, from: null, to: null }
+        : { ...this.state.filteredPipelineState, from: node.id };
+
+    this.setState({ filteredPipelineState: newState });
```

Comment:
this code is being used multiple declare a newState, and then update the filteredState ... maybe we can have a function `updateFilteredPipelineState` that can reduce the redudancy

Reply:
yup good idea, I've created one now. Thank you Rashida 💯